### PR TITLE
Use IIS Implementation for Code Rules, adds UrlRewrite Builder

### PR DESCRIPTION
**Pull request by ZestyBread**
- Instead of using hacky new rule system, we utilize the url rewrite rule structure to implement simple code rules, like RedirectRule and RewriteRule.
- To do this, we create a builder for url rewrite rules, and add matches, conditions, and actions appropriately.
- TODO current the backreference capture is in url rewrite syntax (ex: {R:1} obtains the first back reference in the rule regex match). Do we want this or the standard string format with {1}? I think we should keep the first because again to distinguish between conditions and rules, we need that prefix.
- This is mostly a template for how to implement code rules, more will be added later based on peer and customer feedback.

### CODE REVIEW

**Comment by ZestyBread** [comment]
@Tratcher @davidfowl @muratg

File: src/Microsoft.AspNetCore.Rewrite/CodeRewriteExtensions.cs [hunk 1]
```diff
@@ -35,69 +34,71 @@ public static RewriteOptions AddRules(this RewriteOptions options, List<Rule> ru
...
+        public static RewriteOptions RedirectToHttps(this RewriteOptions options, int statusCode)
+        {
+            return RedirectToHttps(options, statusCode, stopProcessing: false);
+        }
+
+        public static RewriteOptions RedirectToHttps(this RewriteOptions options, int statusCode, bool stopProcessing)
         {
-            options.Rules.Add(new SchemeRule { SSLPort = sslPort, OnCompletion = Transformation.Redirect });
+            var builder = new UrlRewriteRuleBuilder();
+            var rulePattern = InputParser.ParseInputString("https://{HTTP_HOST}/{R:1}");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/CodeRewriteExtensions.cs]: Test what this does for ports.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/CodeRewriteExtensions.cs]: When we get to profiling I bet your prior implementation was more efficient for this simple rule. But we can go back to it if we need to.

File: src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectPathRule.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft.AspNetCore.Rewrite.Internal
+{
+    public class RedirectPathRule : Rule
+    {
+        public Regex MatchPattern { get; set; }
+        public string OnMatch { get; set; }
+        public override RuleResult ApplyRule(RewriteContext context)
+        {
+            return RuleResult.Continue;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectPathRule.cs]: ?

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlRewrite/UrlActions/RedirectAction.cs [hunk 3]
```diff
@@ -30,11 +30,11 @@ public override RuleResult ApplyAction(HttpContext context, MatchResults ruleMat
             {
                 var query =  context.Request.QueryString.Add(new QueryString(pattern.Substring(split)));
                 // not using the response.redirect here because status codes may be 301, 302, 307, 308 
-                context.Response.Headers[HeaderNames.Location] = pattern.Substring(0, split) + query;
+                context.Response.Redirect(pattern.Substring(0, split) + query);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlRewrite/UrlActions/RedirectAction.cs]: Redirect sets the StatusCode

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlRewrite/UrlRewriteFileParser.cs [hunk 4]
```diff
@@ -27,14 +24,14 @@ public static List<UrlRewriteRule> Parse(TextReader reader)
                 var result = new List<UrlRewriteRule>();
                 // TODO Global rules are currently not treated differently than normal rules, fix. 
                 // See: https://github.com/aspnet/BasicMiddleware/issues/59
-                ParseRules(xmlRoot.Descendants(RewriteTags.GlobalRules).FirstOrDefault(), result, isGlobalRule: true);
-                ParseRules(xmlRoot.Descendants(RewriteTags.Rules).FirstOrDefault(), result, isGlobalRule: false);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlRewrite/UrlRewriteFileParser.cs]: how are global rules handled now?

Comment (ZestyBread) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlRewrite/UrlRewriteFileParser.cs]: Not handled for now. See #59. Will address in the future.

File: src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectPathRule.cs (hunk 2 above)
Comment (ZestyBread) [inline src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectPathRule.cs]: deleted

File: src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectToHttpsRule.cs [hunk 5]
```diff
@@ -4,12 +4,12 @@
...
 
-namespace Microsoft.AspNetCore.Rewrite.Internal
+namespace Microsoft.AspNetCore.Rewrite.Internal.CodeRules
 {
-    public class SchemeRule : Rule
+    public class RedirectToHttpsRule : Rule
     {
         public int? SSLPort { get; set; }
-        public Transformation OnCompletion { get; set; } = Transformation.Rewrite;
+        public int StatusCode { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RedirectToHttpsRule.cs]: Unused?

File: src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RewriteToHttpsRule.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+{
+    public class RewriteToHttpsRule : Rule
+    {
+
+        public bool stopProcessing { get; set; }
+        public int? SSLPort { get; set; }
+        public override RuleResult ApplyRule(RewriteContext context)
+        {
+            // TODO this only does http to https, add more features in the future. 
+            if (!context.HttpContext.Request.IsHttps)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/CodeRules/RewriteToHttpsRule.cs]: simplify:
```
if (context.HttpContext.Request.IsHttps)
{
     return RuleResult.Continue;
}
```

**Comment by Tratcher** [comment]
:shipit:

File: BasicMiddleware.sln [hunk 7]
```diff
@@ -5,6 +5,9 @@ MinimumVisualStudioVersion = 10.0.40219.1
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.AspNetCore.HttpOverrides", "src\Microsoft.AspNetCore.HttpOverrides\Microsoft.AspNetCore.HttpOverrides.xproj", "{517308C3-B477-4B01-B461-CAB9C10B6928}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{A5076D28-FA7E-4606-9410-FEDD0D603527}"
+	ProjectSection(SolutionItems) = preProject
```
Comment (ZestyBread) [inline BasicMiddleware.sln]: Why did this get added? @davidfowl

**Pull request merged**